File: training-data/r/linked_list.R
new_node <- function(value, nxt = NULL) {
  list(value = value, nxt = nxt)
}

ll_prepend <- function(head, value) {
  new_node(value, head)
}

ll_append <- function(head, value) {
  if (is.null(head)) return(new_node(value))
  head$nxt <- ll_append(head$nxt, value)
  head
}

ll_to_vector <- function(head) {
  values <- c()
  node <- head
  while (!is.null(node)) {
    values <- c(values, node$value)
    node <- node$nxt
  }
  values
}

list1 <- NULL
list1 <- ll_append(list1, 1)
list1 <- ll_append(list1, 2)
list1 <- ll_append(list1, 3)
list1 <- ll_prepend(list1, 0)
print(ll_to_vector(list1))

File: training-data/fortran/linked_list.f90
module linked_list_mod
    implicit none
    type :: node
        integer :: value
        type(node), pointer :: next => null()
    end type node
contains
    subroutine push_front(head, value)
        type(node), pointer, intent(inout) :: head
        integer, intent(in) :: value
        type(node), pointer :: new_node
        allocate(new_node)
        new_node%value = value
        new_node%next => head
        head => new_node
    end subroutine push_front

    subroutine print_list(head)
        type(node), pointer, intent(in) :: head
        type(node), pointer :: cur
        cur => head
        do while (associated(cur))
            print *, cur%value
            cur => cur%next
        end do
    end subroutine print_list

    function sum_list(head) result(total)
        type(node), pointer, intent(in) :: head
        integer :: total
        type(node), pointer :: cur
        total = 0
        cur => head
        do while (associated(cur))
            total = total + cur%value
            cur => cur%next
        end do
    end function sum_list
end module linked_list_mod

program main
    use linked_list_mod
    implicit none
    type(node), pointer :: head => null()
    call push_front(head, 3)
    call push_front(head, 2)
    call push_front(head, 1)
    call print_list(head)
    print *, sum_list(head)
end program main
